text/0000-default-field-values.md

```diff
@@ -0,0 +1,1979 @@
+
+Default values are only allowed on named fields. There is no syntax provided for
+tuple types like `struct S(i32)` or `enum E { V(i32), }`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Field default values
+
+### Grammar
+
+Let the grammar of record fields in `struct`s and `enum` variants be defined
+like so (in the `.lyg` notation):
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type;
+```
+
+Then, `RecordField` is changed into:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type { "=" def:Expr }?;
+```
+
+Further, given the following partial definition for the expression grammar:
+
+```rust
+Expr = attrs:OuterAttr* kind:ExprKind;
+ExprKind =
+  | ...
+  | Struct:{ path:Path "{" attrs:InnerAttr* fields:StructExprFieldsAndBase "}" }
+  ;
+
+StructExprFieldsAndBase =
+  | Fields:{ fields:StructExprField* % "," ","? }
+  | Base:{ ".." base:Expr }
+  | FieldsAndBase:{ fields:StructExprField+ % "," "," ".." base:Expr }
+  ;
+StructExprField = attrs:OuterAttr* kind:StructExprFieldKind;
+StructExprFieldKind =
+  | Shorthand:IDENT
+  | Explicit:{ field:FieldName ":" expr:Expr }
+  ;
+```
+
+the rule `StructExprFieldsAndBase` is extended with:
+
+```rust
+StructExprFieldsAndBase =| FieldsAndDefault:{ fields:StructExprField+ % "," "," ".." };
+StructExprFieldsAndBase =| Default:{ ".." }
+```
+
+### Static semantics
+
+#### Defining defaults
+
+Given a `RecordField` where the default is specified, i.e.:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type "=" def:Expr;
+```
+
+all the following rules apply when type-checking:
+
+1. The expression `def` must be a constant expression.
+
+2. The expression `def` must coerce to the type `ty`.
+
+3. Generic parameters of the current items are accessible
+   ```rust
+   struct Bar<const A: usize> {
+       field: usize = A,
+   }
+   ```
+
+4. Default const expressions are *not* evaluated at definition time, only
+   during instantiation. This means that the following will not fail to compile:
+   ```rust
+   struct Bar {
+       field1: usize = panic!(),
+       field2: usize = 42,
+   }
+
+   let _ = Bar { field1: 0, .. };
+   ```
+   Having said that, it can be possible to proactivelly attempt to evaluate the
+   default values and emit a lint in a case where the expression is assured to always
+   fail (which would only be possible for expressions that do not reference `const`
+   parameters).
+
+5. The `struct`'s parameters are properly propagated, meaning the following is
+   possible:
+   ```rust
+   struct Bar<T> {
+       field: Vec<T> = Vec::new(),
+   }
+
+   let _ = Bar::<i32> { .. };
+   ```
+
+When lints check attributes such as `#[allow(lint_name)]` are placed on a
+`RecordField`, it also applies to `def` if it exists.
+
+#### Initialization expressions
+
+`Path { fields, .. }` is `const` since the defaulted fields are initialized
+from constants.
```

Comment:
This RFC doesn't mention const promotion anywhere.  That's been on our mind a lot recently.  Perhaps it'd be worth discussing that here.

cc @RalfJung

Reply:
The only question I can think of here is whether `&Path { fields, .. }` can be promoted (assuming all expressions in `fields` are promotable). I think the answer can be yes --that is consistent with `&Path { fields }`, and the main criterion is that the expression never fails to evaluate, which is still given. (This assumes that we get a compiler error when using `Path { field, .. }` and one of the constants in the default values fails to evaluate.)